src/rust/engine/graph/src/lib.rs

```diff
@@ -543,6 +574,28 @@ impl<N: Node> Graph<N> {
     }
   }
 
+  fn detect_cycle(
+    src_id: EntryId,
+    potential_dst_id: EntryId,
+    inner: &mut InnerGraph<N>,
+  ) -> Result<bool, N::Error> {
+    let mut counter = 0;
+    loop {
+      match inner.detect_cycle(src_id, potential_dst_id) {
+        CycleType::Permanent => return Ok(true),
+        CycleType::None => return Ok(false),
+        CycleType::ContainingDirtyNodes(dirty_nodes) => {
+          counter += 1;
+          if counter > 5 {
+            warn!("Couldn't remove cycle containing dirty nodes after {} attempts; nodes in cycle: {:?}", counter, dirty_nodes);
```

Comment:
Although odd, I think that this strategy makes sense. But I think that it will be important to actually render the relevant path (see above re: shortest_path).